feat: persist index credentials in the OS keyring (sysand auth) (#453)
sensmetry/sysand: design/credential-storage.md

  project and every valid token can publish somewhere, so a capability
  boolean is vacuous; per-project authorization stays enforced at the
  upload (its existing `403`). An optional `?project=<id>` pre-flight may
  be added later.

## 7. Publish interaction

Publish is two legs with **different** credential handling:

- **Leg 1, discovery read** (`sysand-index-config.json`, to resolve
  `api_root`): the general read auth policy, unauth-first. `login` scopes
  the credential to cover the discovery/index root (§8), so a private
  index's discovery fetch gets it.
- **Leg 2, upload** (`POST api_root/v1/upload`): **bearer only**, sent
  proactively (an upload cannot be tried unauthenticated then retried).
  Publish checks env bearers first; only when **no env bearer matches**
  does it read the keyring blob (one keychain access). It selects the
  bearer whose glob matches the upload URL.

Publish's bearer selection uses **source precedence**: env bearer matches
first, then keyring, never one flat "exactly one match or error" over a
merged set. Within a source the exactly-one rule stands
(`AmbiguousPublishBearer` is per-source), with candidates carrying the

> identical token collapsing to one match, like the reads rule in section 9. Env and keyring stay **two maps** (source-tagged, with the selected

Reviewer: ```suggestion
identical token collapsing to one match, like the reads rule in section 9.
Env and keyring stay **two maps** (source-tagged, with the selected
```

  tie-breaking; that is only needed once raw-pattern `auth set` or
  same-host nested logins create within-source overlaps (§10).
- **Glob coverage.** `login` anchors the primary glob on the **discovery URL
  the user supplied** (so the discovery fetch itself is authenticated), and
  additionally covers the resolved `index_root` and `api_root` when they
  diverge from it, minimal and non-overlapping. Templated URLs are anchored
  before `{path}` / `{path_raw}`.
- **Glob derivation is escaped and pinned.** URLs can legally contain
  glob metacharacters (an IPv6 literal reads as a globset character
  class; `{path}` templates as alternation), and `GlobMapBuilder` uses
  `literal_separator(true)`. So the derived glob is
  `globset::escape(<normalized root>)` + `**`, with the root normalized
  to end in `/` (for example `https://example.com/idx/**`), and both
  derivation and runtime matching use the same serialization
  (`url::Url::as_str()`) so IDN/percent-encoding agree on both sides. Two
  refinements: a templated root's literal-prefix anchor is **clamped to
  at least `scheme://authority/`** (a query-position placeholder would

>   otherwise degenerate the anchor to `https://**`), skipping the root

Reviewer: I'm confused. Query is after the path, so getting `https://**` with a query-position placeholder should be impossible anyway.

(I agree that requiring at least scheme + authority to be given is sensible, just the wording here is confusing)

Author: (Claude here, replying on Erik's behalf after a brief discussion with him.)

You are right to be confused: the example named the wrong case. A query-position placeholder after any path slash anchors fine (the anchor is the literal prefix cut back to its last `/`, so `https://host/api/files?ref={path}` anchors at `https://host/api/`). The cases the clamp actually guards are a host-position placeholder (`https://{path}.example.com/...`, prefix `https://`) and the corner where the template has no `/` after the authority at all (`https://example.com?file={path}`, where cutting back to the last `/` lands inside `https://`). The check is `template_anchor_root` in `core/src/commands/auth.rs`: cut to last `/`, then require the result to parse as http(s) with a host.

Reworded in 9fb9612b to state that mechanism instead of the misleading example.

Author: (Claude again, after further discussion with Erik.)

Correction to my previous reply: the host-position placeholder never reaches this code at all. `IndexUrlTemplate::parse` (from #440, on main) validates templates by expanding a probe segment that percent-encodes, and since hosts/ports cannot contain `%`, a placeholder in the authority fails URL parsing right there. So the anchor clamp's only reachable degenerate case is a template with no `/` after the authority (`https://example.com?file={path}`), where cutting back to the last `/` lands inside `https://`; the parse check is defense-in-depth for the rest. The doc now says exactly that (de365a82).

  with a notice when even that is not meaningful; and a newly derived

>   root that **subsumes** an already-derived one replaces it, keeping the

Reviewer: I assume subsuming here affects only the current token (i.e. only the URLs of the new login (discovery, index, API)).

Author: (Claude here, replying on Erik's behalf after a brief discussion with him.)

Correct: subsumption operates only within the single login's derived set (the login URL plus its resolved `index_root` and `api_root`, at most three roots). `derive_credential_globs` in `core/src/commands/auth.rs` builds a local roots list seeded with the primary root; other stored records are never touched, and a re-login replaces only its own record. The doc now says this explicitly (9fb9612b).

  set minimal and non-overlapping. Normative test requirements: the
  discovery-document URL, the `index.json` URL, and the upload URL each
  match the compiled derived set (for a template-target login without an
  advertised `api_root`, an encoded-`{path}` project-file URL stands in
  for the upload URL), and an IPv6-literal login (`https://[::1]:8000/`)

>   works.

Reviewer: ```suggestion
  discovery-document URL, the `index.json` URL, and the upload URL
  (if present) each match the compiled derived set, and an IPv6-literal
  login (`https://[::1]:8000/`) works.
```
I think this was a leftover from before requiring explicit `api_root`?

  `SYSAND_CRED_*` lines to set, with the pattern value exact but the
  secret as a **`<token>` placeholder, never the entered value**:
  no-keyring hosts are typically CI/headless where stdout lands in
  captured job logs. Honest posture note: on such hosts the env fallback
  means the secret lives in same-user process environments and typically
  at rest in CI secret config or shell rc files; that is the accepted
  floor there, stated in the docs.
- **Precedence:** `SYSAND_CRED_*` > keyring > unauthenticated (source
  precedence, §8), so CI can override an interactive login. There is no
  separate runtime shadow warning: `auth status` shows per-entry shadowing,
  and the source-named auth-failure messages (§7) identify a stale env var
  exactly when it bites.
- **Expiry:** reactive first: when a request that exercised a stored
  credential ends in **any 4xx** (not just 401, GitLab-style hosts answer
  404 on bad auth) and the record's `expires_at` is past, print
  "credential for `<index>` may be expired or revoked; re-authenticate to
  store a fresh credential". Emitted in the library, so it names no CLI
  command. Proactive when known: a stored `expires_at` lets `auth status`

>   show "expires in N days / expired".

Reviewer: Why does this differ from what publish does (it does not even try to use an (>1 hour) expired token)? Or does "reactive" here apply only to that 1 hour window after expiry, and credential is not tried after that?

Author: Hmmm two things:

- testing use of expired token (up to X minutes)
- reactive use of creds only when needed

The expired/not logic ought to be the same
The reactive use should only be for read access, as that may or may not require a token, but when there is no api access (not that it is explicit), it could be reasonable to always use it even for read situations - at least after being accepted to login (before that, it tries to verify its avtually needed by making a read without first etc)

Reviewer: Thinking more about this, probably not worth checking the expiry before every read; it will be a tiny bit faster for non-expired cases (no expiry checking) and a bit slower for expired cases, but performance does not really matter in that case. Publish on the other hand does quite a bit of work, so checking expiry before doing all that work is worth it.

  configured (no stored credentials, no `SYSAND_CRED_*` variables)." The
  no-usable-keyring note always prints when the backend is unusable (it
  is information, not a negative). No `scheme` column in v1 (always
  bearer for stored; env entries may be basic).
- **Re-login:** `auth login` over an existing entry for the same key
  overwrites it, printing "Replacing existing credential for `<index>`"
  (informational, suppressed under `--quiet`) before the write; the
  previous token is discarded locally (not revoked server-side).

## 10. Scope boundaries

**Deferred to later phases (intended, not v1):**

- `auth set` / `auth unset` (raw-pattern credentials) and the `--pattern`
  override on `login`.
- Basic auth via `--username` (request-time basic via `SYSAND_CRED_*` still
  works).
- Longest-prefix most-specific-glob-wins (needed only once `set` / nested
  logins create within-source overlaps).
  **Out of scope entirely:** acquisition beyond store-what-you-paste (OAuth
  apps, device flows, refresh-token lifecycle); a self-written encrypted vault
  or plaintext credentials file; a user-facing credential "label" concept;

>   multi-account-per-host switching; git credentials (git keeps its own).

Reviewer: ```suggestion

**Out of scope entirely:**

- Acquisition beyond store-what-you-paste (OAuth apps, device flows, refresh-token lifecycle).
- Self-written encrypted vault or plaintext credentials file.
- User-facing credential "label" concept.
- Multi-account-per-host switching.
- git credentials (git keeps its own).
```
formatting

Author: Applied in 4dd94ca5, with your co-author credit. (Claude, on watch duty for Erik.)

## 11. Phases (landed)

The work shipped in phases: the credential store and lazy auth policy
first (a behavior-neutral refactor to soak), then the server-side
`v1/whoami`, then the `auth` commands, then docs. **The whoami-before-login
ordering was load-bearing:** had `login` shipped first, a validating login
against the official index would probe whoami, get a 404, count the API
surface as rejected with the read surface "not tested", and refuse a
valid token.

CI notes that remain true: the workspace test lane runs `sysand-core` and
`sysand` together, and the CLI crate enables `sysand-core/keyring`, so
with resolver 3 feature unification the keyring-gated tests run in the
ordinary test lane (no separate keyring lane); prek's rust hooks are
pinned via `language_version`, since prek does not read

> `rust-toolchain.toml`.

Reviewer: ```suggestion
ordinary test lane (no separate keyring lane).
```
irrelevant

Author: Applied in 4dd94ca5, with your co-author credit. (Claude, on watch duty for Erik.)

## 12. Protocol/spec changes (this repo, `design/`)

`design/index-api-protocol.md` specifies `v1/whoami` (§6), routed under
`api/`. The `design/index-protocol.md` require-`api_root` change landed
separately (an index has an API iff discovery advertises `api_root`; the
plain-URL default was dropped, a breaking change for third-party plain-URL
dynamic indexes, which must now serve `sysand-index-config.json` with an

> explicit `api_root`).

Reviewer: ```suggestion
`design/index-api-protocol.md` specifies `v1/whoami` (§6), routed under
`api/`.
```
out of place here, already explained in `index-protocol.md`

Author: Applied in 4dd94ca5, with your co-author credit. (Claude, on watch duty for Erik.)

| Surface    | Probe path                 | Creds required? |
| ---------- | -------------------------- | --------------- |
| Discovery  | `sysand-index-config.json` | sometimes       |
| Index root | `index.json`               | sometimes       |
| API root   | `v1/whoami`, `v1/upload`   | always          |

Constraints:

- **C1 - unified read access.** Discovery and index root share one auth
  status (public or private together). This removes pathological split
  permutations and collapses the space to a 2x2.
- **C2 - one credential per index.** `auth login` stores one credential per
  index, used for both the read leg and the API leg. Separate read/write
  tokens are not a v1 concept (a later `auth set` could cover them, §10).
- **P2 - API presence is read from discovery (consumed, not enforced
  here).** This plan treats an index as having an API iff its discovery
  document advertises `api_root`, and derives globs accordingly. It does
  **not** change the runtime default (today a plain-URL index still defaults
  `api_root` to its root). Enforcing "`api_root` required" at the protocol
  level, and dropping that default, is a **separate, decoupled change**

>   (§12), out of this plan's phases.

Reviewer: Why require `api_root` to be explicitly advertised to treat it as such?

Author: Without doing so, you could end up not knowing if you failed to access an API or there was no API if the discovery/index was public but not the API, as error codes cant be trusted to distinguish that.

However, I dislike the state of things now with this PR, as we havnt forced sysand publish to require api_root to be explicitly declared.

Wdyt? Make sysand publish require it?

An index must have a file serving index root, but not an API, so it would be good to have the discovery doc (or missing of one) to indicate it - for example with an explicit api_root

Author: Not reflected at great length, but asked ai to make such requirement https://github.com/sensmetry/sysand/pull/478 preliminary

  not given, `login` fails fast ("no terminal for prompt; pass the token
  with `--token-stdin`") instead of hanging or reading a pipe as a secret.
- **Default index.** `sysand auth login` with no URL resolves the target
  from the default-index chain: the `SYSAND_DEFAULT_INDEX` environment
  variable (comma-delimited),
  then a `default = true` index in configuration, else the built-in
  `DEFAULT_INDEX_URL` (`https://sysand.com`). Note `sysand publish` has no
  such default (its `--index` is required), so this chain is defined here,
  not borrowed. If the chain yields **more than one** default index, bare
  `login` errors and asks for an explicit URL. `login` always **echoes the
  resolved index** before prompting (and on the `--token-stdin` path), so a
  project-configured default cannot be targeted silently. `auth status`
  reuses the same chain, but only
  to mark the entries that apply to the default index, and **never
  hard-errors over it**: an ambiguous chain prints a note and marks
  nothing, and a default that is not a valid HTTP(S) index key (for
  example `file://`) silently marks nothing.
- **HTTP(S) only.** `auth login` against a non-HTTP(S) location (for

>   example a local file path, which index resolution accepts elsewhere)

Reviewer: Well, non-http(s) indexes should not be accepted anywhere, even if they currently are.

Author: Incorrect comment, fixed

  errors with "not an HTTP(S) index; nothing to authenticate to".
- **Glob derivation** (§8): automatic from the URL; no manual `--pattern` in
  v1. If derivation is ever wrong for an unusual layout, the `SYSAND_CRED_*`
  env var is the escape hatch until `--pattern` / `auth set` land (§10).
  A **templated URL as the login target itself** (the user typing a
  `{path}` / `{path_raw}` template into `auth login`/`logout`, for example

>   a GitLab repository-files URL) is supported: the storage key is the

Reviewer: I don't understand the context of the first URL mention here: is it from `sysand-index-config.json`, the default index determination, or directly provided by the user? The "templated URL as the login target" seems to cover the directly-provided case, so what is the first case?

Author: Clarified as

> Glob derivation (§8): automatic from the login-target index URL (the URL the user passes to auth login, or the resolved default index when none is given), not from the discovery document; discovery-advertised index_root / api_root add further globs per §8. ...

  template text with its literal-prefix anchor normalized through
  `url::Url` serialization (lowercased scheme and host, default port
  stripped, IDN punycoded; the rest of the template stays verbatim, since
  it is raw text rather than a parsed URL), and the primary glob anchors
  on that literal prefix per §8, exactly like a templated `index_root`
  advertised by discovery. A template with no safe anchor (at least
  `scheme://authority/`) is rejected with a pointer to `SYSAND_CRED_*`.

The index URL is normalized (trailing slash, scheme) before use as the
storage key and for glob derivation, so different spellings do not create
duplicate entries.

## 5. Validation

Login has exactly one behavior, with no opt-out flag or parameter:
discovery is fetched for glob scoping (with the authenticated retry
described below), every surface the index supports is probed, and the
refusal rule decides. A static index has only the read surface; a
dynamic index adds the API.

There is deliberately no `--no-validation` escape (and core's
`do_auth_login` takes no validation parameter, so the bindings expose no

> knob either). Offline or unreachable-index logins already degrade

Reviewer: ```suggestion
There is deliberately no way to disable credential validation against the index
that has an API. Offline or unreachable-index logins already degrade
```

**Probe mechanism.** Validation cannot reuse the runtime unauth-first
policy, which returns only the final response and cannot report whether a
surface actually _accepted_ the credential. Each surface is probed as an
**unauth baseline then a forced-auth retry**: a surface counts as
accepted/tested only when the unauth baseline was a 4xx and the forced retry
then succeeded, so a public surface (200 unauth, credential never sent) is
correctly "not tested", not "accepted". The API surface (`v1/whoami`) is
always authenticated, so its baseline is a known 401 and only the forced
request is needed. Validation is discovery-first (`api_root` is known only
after reading discovery): fetch discovery, resolve `index_root` and
`api_root`, probe `index_root/index.json`, and, **only if discovery
advertised an `api_root`** (not the runtime plain-URL default, §3), probe
`api_root/v1/whoami`, so a static plain-URL index is never phantom-probed
for an API it does not have. Advertised-vs-defaulted needs a flag on the
resolved endpoints (for example `api_root_advertised: bool`), true only in
the explicit-field arm of discovery parsing, false in both defaulting arms

> and the no-document path.

Reviewer: ```suggestion
for an API it does not have.
```
Since API now must be explicitly advertised.
